### ISSUE TITLE
Eureka with AWS ECS Fargate

### ISSUE THREAD

**Issue by dulimitta** [Netflix/eureka#1028]
I am deploying spring cloud application onto AWS ECS.
I updated my eureka client with overriding EurekaInstanceConfigBean as to register with IP address.
It is working well with launch type EC2 but failing with Fargate.
Here is my configuration. I can see null values in the system logs.

``` java
@Bean
	public EurekaInstanceConfigBean eurekaInstanceConfig(InetUtils inetUtils) {
		  EurekaInstanceConfigBean b = new EurekaInstanceConfigBean(inetUtils);
		  AmazonInfo info = AmazonInfo.Builder.newBuilder().autoBuild("eureka");
		  b.setDataCenterInfo(info);
		  info.getMetadata().put(AmazonInfo.MetaDataKey.publicHostname.getName(), info.get(AmazonInfo.MetaDataKey.publicIpv4));
		  System.out.println("*********************Public Host Name*******************************   " + info.get(AmazonInfo.MetaDataKey.publicHostname));
		  System.out.println("*********************Public IPV4 *******************************   " + info.get(AmazonInfo.MetaDataKey.publicIpv4));
		  System.out.println("*********************Local IPV4*******************************   " + info.get(AmazonInfo.MetaDataKey.localIpv4));
		  b.setHostname(info.get(AmazonInfo.MetaDataKey.publicHostname));
		  b.setIpAddress(info.get(AmazonInfo.MetaDataKey.publicIpv4));
		  b.setNonSecurePort(port);
		  return b;
	}

```

**Comment by qiangdavidliu** [comment]
@dulimitta by default eureka client queries the [EC2 metadata endpoint](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ECS_GetStarted.html) for system information. I have not used Fargate yet, but I suspect perhaps this endpoint is not available within Fargate. 

If that is the case, instead of `autoBuild()`, you'll want to manually create the `AmazonInfo` object with system information provided in other ways.

**Comment by dulimitta** [comment]
Sorry for the late reply. How would I get system information in other ways? Can you please suggest.

**Comment by jussiseppala** [comment]
Fargate platform version 1.1 seems to support metadata for ECS. 
See [anouncement](https://aws.amazon.com/about-aws/whats-new/2018/03/aws-fargate-platform-version-1-1/).

I think that eureka-client does not support this.
See [AmazonInfo.java](https://github.com/Netflix/eureka/blob/master/eureka-client/src/main/java/com/netflix/appinfo/AmazonInfo.java)

In case anyone already has implemented this could you please point me to that. Anyway, it would be great to have eureka-client to support ECS (and especially Fargate).

One way to get metadata is to do it as described [here](https://stackoverflow.com/a/50270989/7966995).

**Comment by jussiseppala** [comment]
I created [a sample ](https://github.com/jussiseppala/eurekafargatesample) how to do this. 
Hope this helps!

**Comment by dulimitta** [comment]
Thank you @jussiseppala. I will look into it. We are using ECS with EC2 for now.

**Comment by nitinyadav745** [comment]
@jussiseppala I am running eureka on fargate with public IP disabled for ECS tasks. But, clients are registered on eureka server with with a public IP, which seems to be of underlying Fargate infrastructure. Any help with this would be really helpful

**Comment by Sridharc20** [comment]
Was this case resolved, can somebody post on how to get the container ip and port from the ecs containers (services) and send it to Eureka Server (ecs container).

**Comment by dulimitta** [comment]
@Sridharc20 we are using ECS with EC2. You can refer to @jussiseppala the last comment to get the container IP.

**Comment by mjohns39** [comment]
Can we reopen this please?  I would like to use Eureka on a Fargate cluster and it would be nice if Eureka supported this out of the box without me having to do hackity hacks.

I think I almost have it figured out, I could open a PR once I do.  I am using Spring Cloud though...but my understanding is that configuration for integrating Eureka with Fargate should live here and not as a part of Spring Cloud project.

**Comment by troshko111** [comment]
Reopened, happy to accept a PR with any examples.

**Comment by ChristianEGB** [comment]
@mjohns39 did you made it?

**Comment by shahbaz07** [comment]
@troshko111 I am facing same issue when I register my micro service with euraka it pick wrong ip address. I saw the sample created by @jussiseppala but that will make the code platform specific and I have to replicate that configuration in every micro service. Please let me know how we can fix this issue. If it's already fix please let me know the fix version.

**Comment by ChristianEGB** [comment]
@shahbaz07 did you get it?

**Comment by shahbaz07** [comment]
@ChristianEGB  I didn't check this but I fixed this by adding below code in application class. But if it's fixed already please let me know the fixed version.

> @Bean
    open fun eurekaInstanceConfig(inetUtils: InetUtils?): EurekaInstanceConfigBean? {
        inetUtils?.let {
            val config = EurekaInstanceConfigBean(inetUtils)
            var ip: String? = null
            try {
                ip = InetAddress.getLocalHost().hostAddress
            } catch (e: UnknownHostException) {
                e.printStackTrace()
            }
            config.ipAddress = ip
            config.nonSecurePort = getPort()
            config.isPreferIpAddress = true
            return config
        }
        return null
    }

    open fun getPort(): Int {
        return 0
    }

**Comment by ChristianEGB** [comment]
@shahbaz07 Will check and I will let yoo know, thanks.

**Comment by shahbaz07** [comment]
@ChristianEGB Thanks.

**Comment by shahbaz07** [comment]
@ChristianEGB is there any update on this?

**Comment by anirbandutta9** [comment]
Did anyone figure out how to fix this issue?

**Comment by chrisegb** [comment]
@shahbaz07 @anirbandutta9 in my case I removed Eureka from my infra and I replaced that with the service discovery that AWS ECS already has.

**Comment by mjohns39** [comment]
Yes.  I figured it out.  Apologies, I completely forgot to push a PR.  I will try and do a gist or something at least today...and try to push PR this weekend

**Comment by anirbandutta9** [comment]
> Yes. I figured it out. Apologies, I completely forgot to push a PR. I will try and do a gist or something at least today...and try to push PR this weekend

@mjohns39  Much appreciate ..thanks

**Comment by anirbandutta9** [comment]
> @shahbaz07 @anirbandutta9 in my case I removed Eureka from my infra and I replaced that with the service discovery that AWS ECS already has.

@shahbaz07  We will see if we can implement service discovery in our environment .. Did you get any reference articles on how to do it?

**Comment by mjohns39** [comment]
Here is a gist:  https://gist.github.com/mjohns39/e1f8881620292329b408465a982dc7b8.  I also added the POJOs you'll need.  I want to mention that I grabbed this from somewhere in the wild and just had to change a few things to make it work for me...I will try to find the original author and link to them to give the credit they deserve

**Comment by bhanu-singh** [comment]
Eureka with ECS  https://medium.com/@bhanu.pratap/service-discovery-in-spring-cloud-using-netflix-eureka-in-aws-fargate-2c5c294fee5c
